File: leetcode/Best Time to Buy and Sell Stock.js
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function (prices) {
  const sells = prices.slice(0)

  for (let i = sells.length - 2; i >= 0; --i) {
    sells[i] = Math.max(sells[i], sells[i + 1])
  }

  let profit = 0

  for (let i = 0; i < prices.length; ++i) {
    profit = Math.max(profit, sells[i] - prices[i])
  }

  return profit
}

class Tester {
  constructor(fun) {
    this.fun = fun
    this.total = 0
    this.fail = 0
    this.time = 0
  }

  get pass() {
    return this.total - this.fail
  }

  test(expected, ...input) {
    const start = performance.now()
    const output = this.fun(...input)
    this.time += performance.now() - start

    const isCorrect = output === expected

    console.log('Test', this.total, '...', isCorrect)

    if (!isCorrect) {
      console.error('Output:', output)
      console.error('Expected:', expected)
      ++this.fail
    }

    console.log('')
    ++this.total
  }

  report() {
    console.log(`Passed ${this.pass} / ${this.total} tests
`)
    console.log(`Ran in ${this.time.toFixed(4)}ms`)
  }
}

const t = new Tester(maxProfit)

t.test(5, [7, 1, 5, 3, 6, 4])
t.test(0, [7, 6, 4, 3, 1])
t.test(15, [7, 6, 5, 10, 9, 20])

t.report()

File: leetcode/Best Time to Buy and Sell Stock.py
# https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        i = j = 0
        max_sum = prices[j] - prices[i]
        cur_sum = max_sum

        for k in range(1, n):
            ik_sum = prices[k] - prices[i]
            jk_sum = prices[k] - prices[j]

            if ik_sum > jk_sum:
                j = k
                cur_sum = ik_sum
            else:
                i, j = j, k
                cur_sum = jk_sum

            if cur_sum > max_sum:
                max_sum = cur_sum

        return max(0, max_sum)

    def maxProfit(self, prices: List[int]) -> int:
        min_price = prices[0]
        profit = 0
        for price in prices:
            cur_profit = price - min_price
            if price < min_price:
                min_price = price
            elif profit < cur_profit:
                profit = cur_profit
        return profit


solution = Solution()


def check(inp, expected):
    output = solution.maxProfit(inp)
    assert output == expected


check([7, 1, 5, 3, 6, 4], 5)
check([7, 6, 4, 3, 1], 0)
check([4], 0)
